src/platform/platform_v4.cpp

```diff
@@ -155,13 +155,25 @@ namespace
 
             // Release all other CLR resources
             (void)metahost->Release();
-            (void)runtimeInfo->Release();
 
             // Start the runtime
             hr = runtimeHost->Start();
             IF_FAILURE_RETURN_OR_ABORT(ret, failure_load_runtime, hr, &_prepare_lock);
 
-            (void)runtimeHost->GetCurrentAppDomainId(&_appDomainId);
+            hr = runtimeHost->GetCurrentAppDomainId(&_appDomainId);
+	    if (hr != S_OK)
+            {
+                ICorRuntimeHost* oldRuntimeHost;
+                hr = runtimeInfo->GetInterface(CLSID_CorRuntimeHost, IID_ICorRuntimeHost, (void**)&oldRuntimeHost);
+                IF_FAILURE_RETURN_OR_ABORT(ret, failure_load_runtime, hr, &_prepare_lock);
+                IUnknown* pUnk2;
+                hr = oldRuntimeHost->GetDefaultDomain(&pUnk2);
+                IF_FAILURE_RETURN_OR_ABORT(ret, failure_load_runtime, hr, &_prepare_lock);
+                hr = runtimeHost->GetCurrentAppDomainId(&_appDomainId);
```

Comment:
```suggestion
                hr = runtimeHost->GetCurrentAppDomainId(&_appDomainId);
                (void)oldRuntimeHost->Release();
                (void)pUnk2->Release();
```

@mitchcapper Is the above valid for this scenario? It seems odd that calling `runtimeHost->GetCurrentAppDomainId()` again after the above queries fix the call. Am I missing something here?

Reply:
So after fixing thing we need to still get the app domain of GetDefaultDomainID to know where we should spawn (unless we want to figure out how to get it from the returned domain).    Technically after a failure in theory everything is supposed to be hosed `An unknown catastrophic failure occurred. If a method returns E_FAIL, the CLR is no longer usable within the process. Subsequent calls to hosting methods return HOST_E_CLRNOTAVAILABLE.` however clearly that is not the case but as that is their catch all it is not hard to understand there may be some exceptions.  The other changes requested have been made.

Reply:
@mitchcapper My question is about calling the same function on `runtimeHost` as before the branch into this call. Should we be using `oldRuntimeHost` or `pUnk2` to get the ID instead?

I'm happy if this code works for you, but I'd like to understand if we are relying on some side-effect or if there is something else I am missing.

Reply:
To be clear, I am sure there is a way we could do it on oldRuntimeHost (or probably more appropriately through pUnk2) to get the app ID but I think calling it on the recommended runtimeHost is a good way to go.  This bug is more about properly getting .net to load on a thread and it not being triggered through the current runtime host call.   I couldn't find a better/right way to do this otherwise that wouldn't potentially interfere with an applications own AppDomain setup/calls.   Once we have it loaded the more modern apis work without an issue.  My main goal was to avoid loading as much additional com as possible to keep the impact surface minimal. 

One could get a new instance of the runtime host but in reality I don't think that is somehow better.   The failure code returned from GetCurrentAppDomainId would mean .net is no longer usable within the process for any calls.  It is just their unknown error has happened (and not any actual cleanup killing .net CLR has occurred) and clearly .net is still usable within the process.